internal/server/stat_point.go

```diff
@@ -200,47 +200,113 @@ func (s *Server) GetStatCollection(
-		triples, err := readTriples(ctx, s.store, triplesRowList)
+	// Mapping from stat var to a list of dates. Need to fetch cache data for any
+	// <sv, date> pair.
+	sv2dates := map[string][]string{}
+
+	// If date is not given, get the latest dates and use them to fetch
+	// stats.
+	if date == "" {
+		rowList, keyTokens := buildStatCollectionKeySvObs(parentPlace, childType, "", statVars)
+		dateCache, err := readStatCollection(ctx, s.store, rowList, keyTokens)
 		if err != nil {
 			return nil, err
 		}
-		statVarObject := map[string]*StatisticalVariable{}
-		for statVar, triplesCache := range triples {
-			if triplesCache != nil {
-				statVarObject[statVar], err = triplesToStatsVar(statVar, triplesCache)
-				if err != nil {
-					return nil, err
+		// Two dates are computed here. One is the latest date from all sources and
+		// the other is the date corresponding to the most place entries.
+		for sv, data := range dateCache {
+			latestDate := dateCount{}
+			mostCount := dateCount{}
+			if data != nil {
+				for _, cohort := range data.SourceCohorts {
+					for date, c := range cohort.Val {
+						count := int(c)
+						if count > mostCount.count || (count == mostCount.count && date > mostCount.date) {
+							mostCount = dateCount{date, count}
+						}
+						if date > latestDate.date || (date == latestDate.date && count >= latestDate.count) {
+							latestDate = dateCount{date, count}
+						}
+					}
 				}
 			}
+			// Lastest date is always needed.
+			sv2dates[sv] = []string{latestDate.date}
+			// The date with most place count is needed too.
+			if mostCount.count > latestDate.count {
+				sv2dates[sv] = append(sv2dates[sv], mostCount.date)
+			}
+		}
+	} else {
+		for _, sv := range statVars {
+			sv2dates[sv] = []string{date}
 		}
-		// Construct BigTable row keys.
-		rowList, keyTokens = buildStatCollectionKey(
-			parentPlace, childType, date, statVarObject)
-	}
-	cacheData, err := readStatCollection(ctx, s.store, rowList, keyTokens)
-	if err != nil {
-		return nil, err
 	}
-	for sv, data := range cacheData {
-		if data != nil {
-			cohorts := data.SourceCohorts
-			sort.Sort(SeriesByRank(cohorts))
-			result.Data[sv] = cohorts[0]
+
+	// Get a reverse sorted date list like: [2019, 2017, 2016-01]
+	dateSet := map[string]struct{}{}
+	for _, dates := range sv2dates {
+		for _, date := range dates {
+			dateSet[date] = struct{}{}
 		}
 	}
-	for sv := range result.Data {
-		if result.Data[sv] == nil {
-			result.Data[sv] = &pb.SourceSeries{}
+	dateList := []string{}
+	for date := range dateSet {
+		dateList = append(dateList, date)
+	}
+	sort.Sort(sort.Reverse(sort.StringSlice(dateList)))
+
+	// Query the cache from the most recent date. The order is important as the
+	// source with older date is only used when it has better place coverage.
+	// TODO(shifucun): Parallel the BT query.
+	for _, queryDate := range dateList {
+		rowList, keyTokens := buildStatCollectionKeySvObs(
+			parentPlace, childType, queryDate, statVars)
```

Comment:
At this point, do we need to query all statVars? Is it enough to just query those statVars for which this date exists?